share/spack/gitlab/cloud_pipelines/configs/linux/ci.yaml

```diff
@@ -289,6 +289,6 @@ ci:
         tags: [ "spack", "small" ]
         variables:
           CI_JOB_SIZE: "small"
-          SPACK_BUILD_JOBS: "1"
-          KUBERNETES_CPU_REQUEST: "500m"
-          KUBERNETES_MEMORY_REQUEST: "500M"
+          SPACK_BUILD_JOBS: "2"
+          KUBERNETES_CPU_REQUEST: "2000m"
+          KUBERNETES_MEMORY_REQUEST: "1G"
```

Comment:
You know better than me what the CPU request effectively means. If jobs are scheduled depending on `load + request < #procs`, then you don't need that many extra instances since there's a lot of single threaded IO overhead; if it's based on `total cpu request of all jobs + current request < #procs` then yeah, it needs more instances.

It's not 4x the CPU, there's no such thing of using half a core during builds

Reply:
I think asking for 1 full CPU is probably sufficient. The CPU request is not related to the max loading, but rather the average constant loading if I understand it correctly. Basically, when running with other stuff on the system, how many other things can I also run with this job and not overwhelm the compute resources. Scheduling 8 small jobs running `-j2` on an 8 core node is probably okay.

I think more important is the memory request, because if that is too low the job will die and kill everything else running on the node with it. 1G is probably high, but I think that is fine, better safe than sorry.

Reply:
Can someone figure out what CPU requests mean w.r.t. our particular cluster? Otherwise all we can say is "if certain assumptions, then it's fine, otherwise not". My PR was based on the assumption that the scheduler would consider the current system load, and maybe even current memory usage; it might do something more conservative like taking the sum of cpu/memory requests to be less than what the system provides, so that jobs don't drain the node when they're peaking at the exact same moment.

Reply:
We've always operated under the assumption that these requests (as opposed to limits) are only used by kubernetes when scheduling pods, and do not take into account current node utilization (i.e. kube only looks at the requests made by pods already running on a node, rather than the actual resources getting used).  

My re-reading of the [docs](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#how-pods-with-resource-limits-are-run) seems to confirm that.  To me that sounds like the more conservative option you mentioned.
  
Maybe someone with more knowledge of scheduling on our particular cluster wants to chime in though (@zackgalbreath @AlmightyYakob).

Reply:
Yeah, that's clearly saying that peak matters, not load average:

> The scheduler ensures that, for each resource type, the sum of the resource requests of the scheduled containers is less than the capacity of the node. Note that although actual memory or CPU resource usage on nodes is very low, the scheduler still refuses to place a Pod on a node if the capacity check fails. This protects against a resource shortage on a node when resource usage later increases, for example, during a daily peak in request rate.

Reply:
Hey all! I agree with @haampie, I think it makes sense to allow the jobs access to duel cores to help speed up extraction/compile times. 

Question though, right now we're often provisioning `d3.4xlarge` (16 CPUs, 128GB RAM), `c5.4xlarge` (16 CPUs, 32 GB RAM), and `i3en.3xlarge` (12 CPUs, 96GB RAM) instances. Unless we're over scheduling CPU allocations, that'd seem to mean that we could afford to allocate at least 2GB for every CPU core to make maximal use of memory in the nodes no?

Reply:
The determination for if a node has enough capacity to accommodate the resource requests of a pod is made by the summation of resource requests alone. Here's an excerpt from the docs Scott linked:

> The scheduler ensures that, for each resource type, the sum of the resource requests of the scheduled containers is less than the capacity of the node. Note that although actual memory or CPU resource usage on nodes is very low, the scheduler still refuses to place a Pod on a node if the capacity check fails. This protects against a resource shortage on a node when resource usage later increases, for example, during a daily peak in request rate.

If a limit is defined and a container reaches that limit, the container runtime will step in. If all allocatable resources of a node are reached, pods will begin to be evicted, although I'm not sure in what order/priority that occurs. However, neither of these have any affect on pod scheduling afaik.

Reply:
I guess only data will show if the PR is worth it. Right now nodes are oversubscribed by a factor two in the worst case, which is very big but not causing issues? The current PR would stop oversubscribing nodes, which could result in bad  resource utilization. Alternatively I could set CPU request to 1 and use -j2, and the oversubscription would be the same, although it would still require 2x the number of nodes when many jobs are scheduled 🤷‍♂️

Reply:
Maybe someone who is more informed about actual resource utilization can take over this pr ;p